File: app/domain/relations/derived_ant.py
"""查詢時衍生反義 — 詞林衍生反義與反義端點鏡射（CONTEXT § 詞林衍生反義）。"""

from __future__ import annotations

from typing import List, Optional, Set

from app.domain.relations.graph import CharRelationGraph
from app.domain.relations.ranking import final_score, sort_ant_pool
from app.domain.relations.valid_term import normalize_literal
from app.domain.thesaurus.port import ThesaurusPort

CILIN_DERIVED_SOURCE = "ant_cilin_exanded"
MIRROR_SOURCE = "ant_syn_mirror"
CILIN_DERIVED_CONFIDENCE = 0.75
MIRROR_CONFIDENCE = 0.72


def _pool_literal(text: str) -> Optional[str]:
    return normalize_literal(text)


def _derived_ant_item(
    *,
    char: str,
    source: str,
    confidence: float,
    in_db: bool,
) -> dict:
    return {
        "char": char,
        "relation": "ant",
        "source": source,
        "score": confidence,
        "in_db": in_db,
        "jyutping": "",
        "code": "",
        "_sort": final_score(source=source, confidence=confidence, in_db=in_db),
    }


def runtime_cilin_derived_ant_items(
    query: str,
    seed_chars: List[str],
    *,
    thesaurus: ThesaurusPort,
    present: Set[str],
) -> List[dict]:
    q = query.strip()
    if not q or not seed_chars:
        return []
    out: List[dict] = []
    seen: Set[str] = {q}
    for seed in seed_chars:
        seed = (seed or "").strip()
        if not seed:
            continue
        thesaurus.ensure_loaded()
        for syn in thesaurus.get_cilin_synonyms(seed):
            ch = _pool_literal(syn)
            if not ch or ch == q or ch in seen or ch not in present:
                continue
            seen.add(ch)
            in_db = ch in present
            out.append(
                _derived_ant_item(
                    char=ch,
                    source=CILIN_DERIVED_SOURCE,
                    confidence=CILIN_DERIVED_CONFIDENCE,
                    in_db=in_db,
                )
            )
    return out


def runtime_mirror_ant_items(
    query: str,
    seed_chars: List[str],
    *,
    graph: CharRelationGraph,
    present: Set[str],
    include_static: bool,
) -> List[dict]:
    q = query.strip()
    if not q or not seed_chars:
        return []
    out: List[dict] = []
    seen: Set[str] = {q}
    for seed in seed_chars:
        seed = (seed or "").strip()
        if not seed or seed in seen:
            continue
        seen.add(seed)
        for syn in graph.direct_syn_neighbors(seed, include_static=include_static):
            ch = _pool_literal(syn)
            if not ch or ch == q or ch in seen or ch not in present:
                continue
            seen.add(ch)
            in_db = ch in present
            out.append(
                _derived_ant_item(
                    char=ch,
                    source=MIRROR_SOURCE,
                    confidence=MIRROR_CONFIDENCE,
                    in_db=in_db,
                )
            )
    return out


def append_runtime_derived_ant_pool(
    query: str,
    ant_pool: List[dict],
    *,
    thesaurus: ThesaurusPort,
    graph: CharRelationGraph,
    present: Set[str],
    include_static: bool,
    morpheme_chars: Set[str],
) -> List[dict]:
    seeds = [item.get("char") or "" for item in ant_pool if item.get("char")]
    cilin_items = runtime_cilin_derived_ant_items(
        query, seeds, thesaurus=thesaurus, present=present
    )
    merged = {item["char"]: item for item in ant_pool}
    for item in cilin_items:
        prev = merged.get(item["char"])
        if prev is None or item.get("_sort", 99) < prev.get("_sort", 99):
            merged[item["char"]] = item
    mirror_items = runtime_mirror_ant_items(
        query,
        seeds,
        graph=graph,
        present=present,
        include_static=include_static,
    )
    for item in mirror_items:
        prev = merged.get(item["char"])
        if prev is None or item.get("_sort", 99) < prev.get("_sort", 99):
            merged[item["char"]] = item
    return sort_ant_pool(query, list(merged.values()), morpheme_chars)


__all__ = [
    "CILIN_DERIVED_CONFIDENCE",
    "CILIN_DERIVED_SOURCE",
    "MIRROR_CONFIDENCE",
    "MIRROR_SOURCE",
    "append_runtime_derived_ant_pool",
    "runtime_cilin_derived_ant_items",
    "runtime_mirror_ant_items",
]

File: client/src/db/derived-ant.ts
/**
 * Runtime derived ant — port of derived_ant.py + cilin_derived + mirror_ant (subset)
 * Mirror path uses process-level relation graph (ADR-0050 § 關係圖快取).
 */
import type { Database } from './sqljs.ts';
import { getCilinSynonyms, getStaticAntonyms } from './thesaurus.ts';
import {
  directSynNeighbors,
  ensureRelationGraph,
} from './relation-graph.ts';
import type { RelationPoolItem } from './relation-pool/index.ts';

const CJK_RE = /[\u4e00-\u9fff]/;

const DERIVED_ANT_SOURCES = new Set([
  'ant_syn_mirror',
  'ant_cilin_exanded',
  'ant_syn_bridge',
  'manual_ant_mirror',
]);

export const CILIN_DERIVED_SOURCE = 'ant_cilin_exanded';
export const CILIN_DERIVED_CONFIDENCE = 0.75;
export const MIRROR_SOURCE = 'ant_syn_mirror';
export const MIRROR_CONFIDENCE = 0.72;

function poolLiteral(text: string): string | null {
  const t = (text ?? '')
    .trim()
    .replace(/[（(].*?[）)]/g, '')
    .replace(/\s+/g, '');
  if (!t || t.length > 12 || !CJK_RE.test(t) || /[0-9A-Za-z_]/.test(t)) {
    return null;
  }
  return t;
}

function finalScore(source: string, confidence: number, inDb: boolean): number {
  const ranks: Record<string, number> = {
    guotong: 15,
    project_ant: 12,
    cilin: 10,
    ant_cilin_exanded: 25,
    ant_syn_mirror: 28,
    ant_syn_bridge: 28,
    runtime_static: 80,
    word_relations: 50,
  };
  let rank = 50;
  for (const [key, val] of Object.entries(ranks)) {
    if (source.includes(key)) {
      rank = val;
      break;
    }
  }
  return rank + confidence * 20 + (inDb ? 5 : -10);
}

function derivedAntItem(char: string, source: string, confidence: number, inDb: boolean): RelationPoolItem {
  return {
    char,
    relation: 'ant',
    source,
    score: confidence,
    in_db: inDb,
    jyutping: '',
    code: '',
    group_codes: [],
    _sort: finalScore(source, confidence, inDb),
  };
}

export function directAntSeedsForHead(
  _db: Database,
  head: string,
  membership: Set<string>,
  includeStatic: boolean,
  relAntItems: Array<{ char: string; source: string }>,
): string[] {
  const q = head.trim();
  if (!q) {
    return [];
  }
  const seeds: string[] = [];
  const seen = new Set<string>([q]);
  for (const item of relAntItems) {
    if (DERIVED_ANT_SOURCES.has(item.source || '')) {
      continue;
    }
    const ch = poolLiteral(item.char);
    if (!ch || ch === q || !membership.has(ch) || seen.has(ch)) {
      continue;
    }
    seen.add(ch);
    seeds.push(ch);
  }
  if (includeStatic) {
    for (const ant of getStaticAntonyms(q)) {
      const ch = poolLiteral(ant);
      if (!ch || ch === q || !membership.has(ch) || seen.has(ch)) {
        continue;
      }
      seen.add(ch);
      seeds.push(ch);
    }
  }
  return seeds;
}

function cilinDerivedAntPairs(
  head: string,
  antSeeds: string[],
  membership: Set<string>,
): Array<[string, string]> {
  const h = head.trim();
  if (!h) {
    return [];
  }
  const out: Array<[string, string]> = [];
  const seenTails = new Set([h]);
  for (const seed of antSeeds) {
    const s = seed.trim();
    if (!s) {
      continue;
    }
    for (const syn of getCilinSynonyms(s)) {
      const tail = poolLiteral(syn);
      if (!tail || tail === h || seenTails.has(tail) || !membership.has(tail)) {
        continue;
      }
      seenTails.add(tail);
      out.push([h, tail]);
    }
  }
  return out;
}

function mirrorAntPairs(
  head: string,
  antSeeds: string[],
  synNeighborsOf: (seed: string) => string[],
  membership: Set<string>,
): Array<[string, string]> {
  const h = head.trim();
  if (!h) {
    return [];
  }
  const out: Array<[string, string]> = [];
  const seenTails = new Set([h]);
  for (const seed of antSeeds) {
    const s = seed.trim();
    if (!s) {
      continue;
    }
    for (const syn of synNeighborsOf(s)) {
      const tail = poolLiteral(syn);
      if (!tail || tail === h || seenTails.has(tail) || !membership.has(tail)) {
        continue;
      }
      seenTails.add(tail);
      out.push([h, tail]);
    }
  }
  return out;
}

export async function appendRuntimeDerivedAntPool(
  query: string,
  antPool: RelationPoolItem[],
  db: Database,
  membership: Set<string>,
  includeStatic: boolean,
  _morphemeChars: Set<string>,
  headSyns: Set<string>,
  relAntItems: Array<{ char: string; source: string }>,
): Promise<RelationPoolItem[]> {
  const seeds = directAntSeedsForHead(db, query, membership, includeStatic, relAntItems);
  const present = membership;
  const merged = new Map(antPool.map((item) => [item.char, item]));

  for (const [, tail] of cilinDerivedAntPairs(query, seeds, present)) {
    if (!tail || headSyns.has(tail)) {
      continue;
    }
    const item = derivedAntItem(tail, CILIN_DERIVED_SOURCE, CILIN_DERIVED_CONFIDENCE, present.has(tail));
    const prev = merged.get(tail);
    if (!prev || item._sort < prev._sort) {
      merged.set(tail, item);
    }
  }

  await ensureRelationGraph(db, membership, includeStatic);
  for (const [, tail] of mirrorAntPairs(query, seeds, (s) => directSynNeighbors(s), present)) {
    if (!tail || headSyns.has(tail)) {
      continue;
    }
    const item = derivedAntItem(tail, MIRROR_SOURCE, MIRROR_CONFIDENCE, present.has(tail));
    const prev = merged.get(tail);
    if (!prev || item._sort < prev._sort) {
      merged.set(tail, item);
    }
  }

  return [...merged.values()];
}

/** ponytail: runnable self-check — `npx tsx client/scripts/derived-ant-self-check.ts` */
export function derivedAntLogicSelfCheck(db: Database): void {
  const membership = new Set(['快樂', '悲傷', '傷心', '哀愁']);
  const relAnt = [{ char: '悲傷', source: 'guotong' }];
  const seeds = directAntSeedsForHead(db, '快樂', membership, true, relAnt);
  if (!seeds.includes('悲傷')) {
    throw new Error('derivedAntLogicSelfCheck: seeds');
  }
  const cilinPairs = cilinDerivedAntPairs('快樂', seeds, membership);
  if (!cilinPairs.some(([, t]) => t === '傷心')) {
    throw new Error('derivedAntLogicSelfCheck: cilin');
  }
}
